---
title: "Arkansas Field Data 2016"
output: html_notebook
---

## Load in Data
Here we will import all of the necessary files. I will first import the counts table. This table was converted from the original biom file to a tab separated file. Next is the mapping file, which is the metadata. After that is the taxonomy file, which contains all of the taxonomies for the OTUs in the count table. The last file contains all of the OTUs that have been assigned as either plastid or mitochondria. 
```{r}
library(lubridate)
library(tidyverse)
setwd("~/Google Drive/RMB/Analyses/ArkansasField/ARK16/FULLData/Data/")
counts <- read.table(gzfile("ark16_otutable.txt.gz"), header = T, row.names = 1, sep = "\t")
map <- read.table("../Data/ark16.map", header = T, sep = "\t") %>% mutate(Date = mdy(Date)) %>% mutate(Days = as.numeric(Date - min(Date)))
tax <- readRDS("gg_otus_tax.rds")
organelle_otus <- readRDS("organelle.rds")
```

## Remove organellar reads from the data
Mitochondria and plastids have 16S rRNA that gets sequenced along with the all of the other bacteria. These need to be removed, as they are not part of the microbiome.
```{r}
countsNoOrg <- counts[!row.names(counts)%in%organelle_otus,]
```

## Make sure the mapping file is in the same order as the counts
I will also add on sampling depth as a column in the map. If a sample has 200 reads or so, I usually throw it away. But the minumum here is a little bit less than 4000, so I think it's okay.
```{r}
map <- map[match(colnames(countsNoOrg), map$SampleID),]
map$Depth <- colSums(countsNoOrg)
map %>% 
  ggplot(., aes(Depth, fill = Compartment)) +
  geom_histogram() +
  facet_grid(.~Compartment)
```

## Normalize the counts table
There are a few ways of normalizing counts. I will employ using relative abundance here. It's pretty much just the looking at the percent of each microbe in each sample
```{r}
RA <- t(t(countsNoOrg) / colSums(countsNoOrg)) * 1000
```

## Remove OTUs that do not show up in 5% of the samples
I will remove microbes that do not show up regularly. These are noisy OTUs, and while they may be real, they most likely popped up through contamination, sequencing and PCR error. They make the downstream stats noisy, and so I will remove them.
```{r}
occurence <- function(x){
  return(length(x[x > 0]))
}
RA_clean <- RA[apply(RA, 1, occurence) >= ncol(RA) * 0.05,]
cat(nrow(RA), "OTUs in original table.\n")
cat(nrow(RA_clean), "OTUs in cleaned table.\n")
```

## Make a tidy version of the data
```{r}
library(reshape2)
ark.data <- melt(cbind(map, t(RA_clean)), id.vars = names(map)) %>% 
  inner_join(tax, by = "variable")
```


# Statistics
## Permanova
Permanova will tell you if an experimental factor is important for describing differences between your samples of multivariate data.
```{r}
library(vegan)
adonis(t(log2(RA_clean + 1)) ~ Compartment + Days * Cultivar, strata = factor(map$Lane), data = map)
```
This is telling us that Compartment, Days, and Cultivar are all significant factors explaining differences between the microbiomes of our samples. There is also a significant interaction of Cultivar with Days suggesting that the differences between the compartments is dependent on the age of the plant. The relative effect size is given by the 'R2' value.

So now that we know that some of these factors are important, let's go ahead and visualize the clustering patterns with PCoA.
```{r}
ark.pc <- capscale(t(log2(RA_clean + 1)) ~ 1)
ark.pc_axes <- cbind(map, scores(ark.pc, choices = c(1:3))$sites)
ark.pc_axes %>% 
  ggplot(., aes(MDS1, MDS2, color = factor(Date), shape = Compartment)) +
  geom_point(size = 2) +
  theme_minimal() +
  labs(color = "Date")
```

## Distribution of Phyla 
You can play around with the different levels that you would like to look like. This code is finding the top 10 most abundant phyla (Proteobacteria is broken down to it's various classes because it's a super phylum) and plotting them out.
```{r}
ark.data %>% 
  group_by(Phylum2) %>% 
  summarise(total = sum(value)) %>% 
  arrange(-total) %>% 
  head(10) %>% 
  inner_join(ark.data, by = "Phylum2") %>% 
  group_by(SampleID, Compartment, Cultivar, Plot, Phylum2, Date, Rep) %>% 
  summarise(total = sum(value)) %>% 
  ggplot(., aes(x = paste(Plot, Rep), y = total, fill = Phylum2))+
  geom_bar(stat = "identity") +
  labs(x = "Individual Samples", y = "Relative Abundance") +
  facet_grid(Compartment ~ Date, scales = "free") +
  scale_fill_brewer(palette = "Spectral") +
  theme_minimal() +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), legend.position = "bottom")
```

### Make a function to plot OTU abundance
```{r}
otu.plot <- function(., method = "lm"){
  ggplot(., aes(Days, value, color = Cultivar)) +
    geom_point() +
    geom_smooth(method = method, se = F) +
    facet_grid(.~Compartment) +
    scale_color_brewer(palette = "Accent")
}
```

I'm going to throw this section out eventually.
```{r}
comp.models <- ark.data %>% 
  group_by(variable) %>% 
  summarise(sum = sum(value)) %>% 
  arrange(-sum) %>% 
  head(4000) %>% 
  inner_join(ark.data, by = "variable") %>% 
  group_by(variable) %>% 
  nest() %>% 
  mutate(models = map(data, ~lm(log2(value + 1) ~ Compartment*Days, .))) %>% 
  mutate(ANOVA = map(models, ~anova(.)))

comp.results <- comp.models %>% 
  unnest(map(ANOVA, tidy))

comp.results %>% 
  filter(term == "Compartment:Days") %>% 
  mutate(p.adj = p.adjust(p.value)) %>% 
  inner_join(tax, by = "variable") %>% 
  filter(p.adj <= 0.01) %>% 
  arrange(-p.adj) %>% 
  filter(grepl("Methyl", Family))

cult.models <- ark.data %>% 
  group_by(variable) %>% 
  summarise(sum = sum(value)) %>% 
  arrange(-sum) %>% 
  head(4000) %>% 
  inner_join(ark.data, by = "variable") %>% 
  filter(Compartment != "Bulk Soil") %>% 
  group_by(variable, Compartment) %>% 
  nest() %>% 
  mutate(models = map(data, ~rlm(log2(value + 1) ~ Cultivar*Days, .))) %>% 
  mutate(ANOVA = map(models, ~anova(.)))

cult.anova.results <- cult.models %>% 
  unnest(map(ANOVA, tidy))
cult.lm.results <- cult.models %>% 
  unnest(map(models, tidy))

cult.anova.results %>% 
  inner_join(tax, by = "variable") %>% 
  group_by(term, Compartment, Phylum2) %>% 
  mutate(p.adj = p.adjust(p.value)) %>% 
  filter(term == "Cultivar:Days") %>% 
  filter(p.adj <= 0.05) %>% 
  ungroup() %>% 
  dplyr::select(variable, Compartment) %>% 
  inner_join(cult.lm.results, by = c("variable", "Compartment")) %>% 
  filter(term == "Days" | term == "CultivarSabine:Days") %>% 
  mutate(term = gsub("CultivarSabine:Days", "CultivarSabineDays", term)) %>% 
  dplyr::select(variable, Compartment, term, estimate) %>% 
  spread(term, estimate) %>% 
  rename(CLXLSlope = Days) %>% 
  mutate(SabineSlope = CLXLSlope + CultivarSabineDays) %>% 
  arrange(CultivarSabineDays) %>% 
  ggplot(., aes(fct_reorder(variable, CultivarSabineDays), CultivarSabineDays)) +
  geom_point() +
  facet_wrap(~Compartment, scales = "free")

```
```{r}
ark.data %>% 
  filter(variable == "533") %>% 
  otu.plot(method = "lm")
```



## Doing stats with negative binomial GLM
First requires that we have count data that is normalized. We will use rarefaction for this job. I hate this method because we have to throw away so much of the data. The minimum sequencing depth was 4384 sequences. We will randomly sample 4384 sequences from each sample.
```{r}
rare <- t(rrarefy(t(countsNoOrg), sample = min(colSums(countsNoOrg))))
rare.clean <- rare[apply(rare, 1, occurence) >= ncol(rare) * 0.05,]
rare.clean <- rare.clean[,match(map$SampleID, colnames(rare.clean))]
cat(nrow(rare), "OTUs in original table.\n")
cat(nrow(rare.clean), "OTUs in cleaned table.\n")
```

Just for fun I will also do the beta diversity stats here on this new dataset.
```{r}
adonis(t(log2(rare.clean + 1)) ~ Compartment + Days*Cultivar, strata = factor(map$Lane), data = map)
```
```{r}
ark.pc_rare <- capscale(t(log2(rare.clean + 1)) ~ 1)
ark.pc_axes_rare <- cbind(map, scores(ark.pc_rare, choices = c(1:3))$sites)
ark.pc_axes_rare %>% 
  ggplot(., aes(MDS1, MDS2, color = factor(Date), shape = Compartment)) +
  geom_point(size = 2) +
  theme_minimal() +
  labs(color = "Date")
```

Here we will convert the data to the 'long' format.
```{r}
rare.data <- melt(cbind(map, t(rare.clean)), id.vars = names(map)) %>% inner_join(tax, by = "variable")
```

## Modelling effects using a negative binomial glm. This is where the stats get real. The linear models employed above are theoretically inappropriate for this data type. Here we have count data which effects the way we should approach modelling. We could use a glm with a poisson distribution family, but our data is overdispersed. This means that the variances of the OTUs are greater than the means. Let's take a look at this.

```{r}
rare.data %>% 
  group_by(variable) %>% 
  summarise(mean = mean(value), var = var(value)) %>% 
  ggplot(aes(mean, var)) +
  geom_point(alpha = 0.1) +
  scale_x_log10() +
  scale_y_log10() +
  labs(x = "Mean Abundance", y = "Variance")
```

This means that poisson won't model our counts well and we should look at another family of regression called negative binomial (NB). NB models can handle overdispersed data. We can run a NB GLM through the MASS package. Let's go ahead and load the required packages.
```{r}
library(MASS)
library(stringr)
library(forcats)
library(broom)
```
The strategy will be to run a glm for each OTU. When you run models like this, there is a possibility that one of the models will fail and the whole data frame will not get saved because R throws an error. We will use the possibly() adverb from the purrr package to get around this. If there is an error, the possibly adverb tells the map function to just fill the output with an NA.
```{r}
safe_glm <- possibly(glm.nb, NA_real_)
```
Let's first find OTUs that are significantly different across compartments and over time. The baseline will be bulk soil.
```{r}
comp.glm <- rare.data %>% 
  group_by(variable) %>% 
  summarise(prev = sum(value > 0) / n()) %>% 
  filter(prev >= 0.1) %>% 
  inner_join(rare.data, by = "variable") %>% 
  group_by(variable) %>% 
  na.omit() %>% 
  nest() %>% 
  mutate(nb_model = map(data, ~safe_glm(value ~ Compartment*Days, .))) %>% 
  mutate(type = "No offset")

comp.glm_results <- comp.glm %>%
  unnest(map(nb_model, tidy)) %>% 
  inner_join(tax, by = "variable") %>% 
  mutate(p.adj = p.adjust(p.value, "BY"))

comp.glm_results %>% 
  filter(p.adj <= 0.01) %>%
  arrange(estimate)
  filter(term == "CompartmentEndosphere:Days" | term == "CompartmentRhizosphere:Days") %>% 
  ggplot(aes(fct_reorder(variable, estimate), estimate, color = Phylum2)) +
  geom_point() +
  facet_grid(.~term)
```

```{r}
cult.glm <- rare.data %>% 
  group_by(variable, Compartment) %>% 
  summarise(prev = sum(value > 0) / n()) %>% 
  filter(prev >= 0.1) %>% 
  inner_join(rare.data, by = c("variable", "Compartment")) %>% 
  filter(Compartment != "Bulk Soil") %>% 
  group_by(Compartment, variable) %>% 
  na.omit() %>% 
  nest() %>% 
  mutate(nb_model = map(data, ~safe_glm(value ~ Cultivar*Days, .))) %>% 
  mutate(type = "No offset")
cult.glm %>% 
  head()
```

How do we investigate the results of the multiple models? It would be hell to run a summary on each model to gather the coefficients. It would also be a pain to write your own function to do this. Fortunately the Broom package offers a concise and intuitive function to do this using tidy(). Because we have a nice nested dataframe, we can input this into map and get the coefficients for each of the models.
```{r}
cult.glm_results <- cult.glm %>%
  inner_join(tax, by = "variable") %>% 
  unnest(map(nb_model, tidy)) %>%
  mutate(p.adj = p.adjust(p.value, method = "BY"))
```

There are several ways of going deeper into the models to find trends. Quantitative geneticists rely on manhattan plots to identify loci in the genome associated with certain traits. On the x-axis, these plots show the genomic locus. On the y-axis, these plots show some kind of score for a correlative measure. The larger the peak, the stronger the correlation. Here, I've made a modified version of this technique suited to microbes. Of course there are no genomes to work with here. Instead, we can group the microbes by the phylogenetic relationships. One could use the topology of the phylogenetic tree of 16S rRNA genes to do this. I'm not going to do that here because there are all kinds of polytomies in the tree, so it can get messy. Instead I will group them by their Phylum, Order, and Families. So for instance, all the Protoebacteria will be together sorted by their orders and families. All of the Actinobacteria will be together sorted by their orders and families. Even though the topology of the tree is messed up, the order still follows a heirarchical pattern.

I have modified the y-axis here to represent whether a microbe is up or down depending on the comparison. For instance, if the comparison is Days. A significant value < 0 represents microbes that were significantly decreasing over the season. Above 0 reprepsents microbes that were significantly increasing over the season. If the comparison is cultivar, then above 0 will be OTUs enriched in Sabine and below 0 will be OTUs enriched in CLXL745.
```{r}
manhattan_plot <- function(model_results, model_term) {
  plot.dat <- model_results %>% 
    filter(term == model_term) %>% 
    arrange(paste(Phylum, Phylum2, Class, Order, Family)) %>% 
    mutate(rank = 1:nrow(.)) %>% 
    mutate(direction = ifelse(estimate > 0, 1, -1)) %>% 
    mutate(sig = ifelse(p.adj <= 0.05, "sig", "ns")) %>% 
    mutate(variable = fct_reorder(variable, rank))

  areas <- plot.dat %>% 
    group_by(Phylum2) %>% 
    summarise(min = min(rank), max = max(rank)) %>% 
    mutate(type = 1:n()) %>% 
    mutate(type2 = ifelse(type %% 2 == 0, "even", "odd"))
  
  ggplot(plot.dat) +
    geom_rect(data = areas, aes(xmin = min, xmax = max, ymin = -Inf, ymax = Inf, fill = type2), alpha = 0.3) +
    geom_point(data = plot.dat %>% filter(sig == "ns"), aes(rank, -log10(p.adj) * direction, size = abs(estimate)), color = "black", alpha = 0.01) +
    geom_point(data = plot.dat %>% filter(sig == "sig"), aes(rank, -log10(p.adj) * direction, color = Family, size = abs(estimate), alpha = abs(statistic))) +
    scale_size(range = c(0,15)) +
    facet_grid(Compartment ~ .) +
    theme_bw() + 
    labs(x = "Microbe", y = "-Log10(P)") +
    theme(panel.grid.major=element_blank(),panel.grid.minor=element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank()) +
    scale_fill_manual(values = c("grey70", "white"))
}
```

```{r}
cult.glm_results %>% 
  manhattan_plot(., "CultivarSabine:Days")
```


```{r}
rare.data %>% 
  filter(variable == "1139084") %>% 
  otu.plot()
rare.data %>% 
  ggplot(aes(log(value + 1))) +
  geom_histogram()
```



